Allow callable instance scope to throw exceptions

kanstantsin-chernik opened the pull request:
Details:
Allow callable instance scope to throw exceptions

Test plan: covered with functional tests

Reviwers:
@kris-jusiak @krzysztof-jusiak

kanstantsin-chernik committed:
Allow callable instance scope to throw exceptions

kris-jusiak commented:
Thanks. It's good stuff @chenar. 
Thanks for the description, tests and di.hpp update 👍 
I think that the only thing left would be to update spec and get rid of noexcept when appropirate (http://boost-experimental.github.io/di/user_guide/index.html).

I have no issues with your changes but let me fix the master CI first so that DI will be green again before merging.

Commit:
Remove noexcept from factory

Commit:
Added notice for dynamic binding of exception support

kanstantsin-chernik commented:
@krzysztof-jusiak Thank you for feedback. I've updated factory extension and added `notice` for dynamic binding section. 
If you have any ideas or comments to wording let me know I will improve that. #239

kris-jusiak commented on test/ft/di_injector_except.cpp:
```diff
@@ -26,3 +26,80 @@ test should_throw_when_ctor_throws = [] {
+
+test should_throw_when_factory_with_1_arg_throws = [] {
+	struct except_factory {
+		template <class TInjector>
```
Templates cannot be declared inside of a local class, so that won't compile in clang/gcc. Does it work in msvc?

kanstantsin-chernik replied:
@krzysztof-jusiak Yes, it perfectly works for MSVC.
Thank you for info. I will fix it in a minute

kris-jusiak committed:
Merge branch 'cpp14' into callable_exceptions

Commit:
Move template out of lambda
